Added kl_divergence for multivariate normals

lumip opened the pull request:
KL implementation for `MultivariateNormal`

EDIT: just saw that there is already a pull request for this from a year ago (https://github.com/pyro-ppl/numpyro/pull/1487). What's the hold up with that?

fehiepsi commented on numpyro/distributions/kl.py:
```diff
@@ -134,6 +136,66 @@ def kl_divergence(p, q):
+            f" {p.event_shape} and {q.event_shape} for p and q, respectively."
+        )
+
+    if p.batch_shape != q.batch_shape:
```
This check is unnecessary. It looks to me that the only operator which does not support batching is `solve_triangular`. I'll comment below.

fehiepsi commented on numpyro/distributions/kl.py:
```diff
@@ -134,6 +136,66 @@ def kl_divergence(p, q):
+        ))
+
+        z = jnp.matmul(Lq_inv, (p_mean - q_mean))
+        t1 = jnp.dot(z, z)
```
You can use [_batch_mahalanobis](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/distributions/continuous.py#L1370) for this:
```
t1 = _batch_mahalanobis(q.scale_tril, q.loc - p.loc)
```

fehiepsi commented on numpyro/distributions/kl.py:
```diff
@@ -134,6 +136,66 @@ def kl_divergence(p, q):
+
+        tr = jnp.sum(jnp.diagonal(
+            Lq_inv.T @ (Lq_inv @ p_scale_tril) @ p_scale_tril.T
+        ))
```
You can use [_batch_solve_triangular](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/distributions/continuous.py#L1240) for this:
```
tmp = _batch_solve_triangular(q.scale_tril, p.scale_tril)
tr = jnp.square(tmp).sum((-1, -2))
```

fehiepsi replied:
Actually, I like your approach better. We can
```Lq_inv = vmap(lambda A, x: solve_triangular(A, x, lower=True))(q.scale_tril.reshape((-1,) + q.scale_tril.shape[-2:]), jnp.eye(D))```.

Then compute
```
tr = jnp.square(Lq_inv @ p.scale_tril).sum((-1, -2))
t1 = jnp.square(Lq_inv @ (p.loc - q.loc)[..., None]).sum((-1, -2))
```

fehiepsi reviewed (commented):
Thank you @lumip! The author of #1487 did not response so I think we can use this PR instead. I have a few comments below to address the batching issues.

lumip commented:
Thanks for the feedback @fehiepsi . I have modified the code according to your suggestions and also added a test case where the batch dimensions of the two distributions are not identical (and fixed the linter complaints). If everything is good now, I'd still rebase onto the recent master and merge the commits.

fehiepsi commented on numpyro/distributions/kl.py:
```diff
@@ -142,58 +143,45 @@ def kl_divergence(p: MultivariateNormal, q: MultivariateNormal):
 
-        Lq_inv = solve_triangular(q_scale_tril, jnp.eye(D), lower=True)
+    q_half_log_det = jnp.log(jnp.diagonal(q.scale_tril, axis1=-2, axis2=-1)).sum(-1)
+    assert q_half_log_det.shape == q.batch_shape
```
nit: this assertion might not be true. In MultivariateNormal implementation, we avoid unnecessary broadcasting (e.g. we can have a batch of means with a single scale_tril).

lumip replied:
Right, I wasn't thinking of that. Removed the assertion and added tests for those cases.

fehiepsi commented on numpyro/distributions/kl.py:
```diff
@@ -142,58 +143,45 @@ def kl_divergence(p: MultivariateNormal, q: MultivariateNormal):
 
-    if p.batch_shape != q.batch_shape:
+    min_batch_ndim = min(len(p.batch_shape), len(q.batch_shape))
+    if p.batch_shape[-min_batch_ndim:] != q.batch_shape[-min_batch_ndim:]:
```
how about only assert that p.batch_shape and q.batch_shape can be broadcasted.
```
try:
    result_batch_shape = jnp.broadcast_shapes(p.batch_shape, q.batch_shape)
except ValueError:
    raise ...
```

lumip replied:
done

fehiepsi commented on numpyro/distributions/kl.py:
```diff
@@ -142,58 +143,45 @@ def kl_divergence(p: MultivariateNormal, q: MultivariateNormal):
-        assert jnp.ndim(q_mean) == 1
-        assert jnp.ndim(p_scale_tril) == 2
-        assert jnp.ndim(q_scale_tril) == 2
+    assert q.mean.shape == q.batch_shape + q.event_shape
```
those assertions are unnecessary.

lumip replied:
removed

fehiepsi commented on numpyro/distributions/kl.py:
```diff
@@ -142,58 +143,45 @@ def kl_divergence(p: MultivariateNormal, q: MultivariateNormal):
 
-        return .5 * (tr + t1 - D - log_det_ratio)
+    tr = _batch_trace_from_cholesky(Lq_inv @ p.scale_tril)
+    assert tr.shape == result_batch_shape
```
this assertion might not be true.

lumip replied:
fixed

fehiepsi commented on numpyro/distributions/kl.py:
```diff
@@ -142,58 +143,45 @@ def kl_divergence(p: MultivariateNormal, q: MultivariateNormal):
-    p_mean_flat = jnp.reshape(p.mean, (-1, D))
-    p_scale_tril_flat = jnp.reshape(p.scale_tril, (-1, D, D))
+    t1 = jnp.square(Lq_inv @ (p.loc - q.loc)[..., jnp.newaxis]).sum((-2, -1))
+    assert t1.shape == result_batch_shape
```
this assertion might not be true

lumip replied:
fixed

fehiepsi commented on numpyro/distributions/kl.py:
```diff
@@ -142,58 +143,45 @@ def kl_divergence(p: MultivariateNormal, q: MultivariateNormal):
-        ).sum(-1)
-        log_det_ratio = 2 * (p_half_log_det - q_half_log_det)
+    p_half_log_det = jnp.log(jnp.diagonal(p.scale_tril, axis1=-2, axis2=-1)).sum(-1)
+    assert p_half_log_det.shape == p.batch_shape
```
this assertion might not be true

lumip replied:
removed

fehiepsi commented on test/test_distributions.py:
```diff
@@ -2849,12 +2849,24 @@ def test_kl_expanded_normal(batch_shape, event_shape):
+        ((), ()),
+        ((1,), (1,)),
+        ((2, 3), (2, 3)),
+        ((5, 2, 3), (2, 3)),
```
could you change this to `(5, 1, 3)` and `(2, 3)`?

lumip replied:
done

fehiepsi commented on test/test_distributions.py:
```diff
@@ -2849,12 +2849,24 @@ def test_kl_expanded_normal(batch_shape, event_shape):
+        ((1,), (1,)),
+        ((2, 3), (2, 3)),
+        ((5, 2, 3), (2, 3)),
+        ((2, 3), (5, 2, 3)),
```
nit: maybe `((1, 3), (5, 2, 3))`?

lumip replied:
done

fehiepsi reviewed (commented):
LGTM, thanks @lumip! I have a few small comments on the assertions.

fehiepsi reviewed (approved):
Thanks, @lumip!
